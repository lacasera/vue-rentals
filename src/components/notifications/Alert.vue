<template>
  <div :class="alertClass" role="alert" >
    <div class="flex">
      <div class="py-1">
        <svg :class="iconClass" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
          <path
            d="M2.93 17.07A10 10 0 1 1 17.07 2.93 10 10 0 0 1 2.93 17.07zm12.73-1.41A8 8 0 1 0 4.34 4.34a8 8 0 0 0 11.32 11.32zM9 11V9h2v6H9v-4zm0-6h2v2H9V5z"
          />
        </svg>
      </div>
      <div>
        <p id="alert-title" class="font-bold"> {{ title}} </p>
        <p id="alert-message" class="text-sm"> {{message}}</p>
      </div>
    </div>
  </div>
</template>
<script>
import { mapMutations } from 'vuex'

export default {
  name: 'Alert',
  mounted () {
    setTimeout(() => {
      this.setNotification({
        show: false,
        type: '',
        title: '',
        message: ''
      })
    }, 5000)
  },
  methods: {
    ...mapMutations(['setNotification'])
  },
  data () {
    return {
      classMap: {
        error: {
          header: 'bg-red-100 border-red-500 text-red-900',
          icon: 'text-red-500'
        },
        success: {
          header: 'bg-teal-100 border-teal-500 text-teal-900',
          icon: 'text-teal-500'
        },
        warning: {
          header: 'bg-orange-100 border-orange-500 text-orange-900',
          icon: 'text-orange-500'
        },
        info: {
          header: 'bg-blue-100 border-blue-500 text-blue-900',
          icon: 'text-blue-500'
        }
      }
    }
  },
  computed: {
    alertClass: function () {
      return this.type ? 'border-t-4 rounded-b px-4 py-3 shadow-md '.concat(this.classMap[this.type]['header']) : ''
    },
    iconClass: function () {
      return this.type ? 'fill-current h-6 w-6 text-red-500 mr-4 '.concat(this.classMap[this.type]['icon']) : ''
    }
  },
  props: {
    title: {
      type: String,
      required: true
    },
    type: {
      type: String,
      required: true,
      validator: function (value) {
        return ['success', 'info', 'warning', 'error'].indexOf(value) !== -1
      }
    },
    message: {
      type: String,
      required: true
    }
  }
}
</script>
