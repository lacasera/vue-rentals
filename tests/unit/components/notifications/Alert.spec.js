import { shallowMount, mount } from '@vue/test-utils'
import Alert from '@/components/notifications/Alert'

const spy = jest.spyOn(console, 'error')

describe('Alert', () => {
  afterEach(() => spy.mockReset())

  it('should render properly', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'error',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with the text passed', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'error',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.find('div').text()).toEqual('Hello   Word')
  })

  it('should throw error if "type" prop is not passed', () => {
    mount(Alert, {
      propsData: {
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "type"'))
  })

  it('should throw error if "title" prop is not passed', () => {
    mount(Alert, {
      propsData: {
        type: 'error',
        message: 'Word'
      }
    })

    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "title"'))
  })

  it('should throw error if "message" prop is not passed', () => {
    mount(Alert, {
      propsData: {
        type: 'error',
        title: 'Word'
      }
    })

    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "message"'))
  })

  it('should have the right success class', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'error',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.classes()).toEqual([
      'border-t-4',
      'rounded-b',
      'px-4',
      'py-3',
      'shadow-md',
      'bg-red-100',
      'border-red-500',
      'text-red-900'
    ])
  })

  it('shoudl have the right error class', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'success',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.classes()).toEqual([
      'border-t-4',
      'rounded-b',
      'px-4',
      'py-3',
      'shadow-md',
      'bg-teal-100',
      'border-teal-500',
      'text-teal-900'
    ])
  })

  it('shoudl have the right info class', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'info',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.classes()).toEqual([
      'border-t-4',
      'rounded-b',
      'px-4',
      'py-3',
      'shadow-md',
      'bg-blue-100',
      'border-blue-500',
      'text-blue-900'
    ])
  })

  it('shoudl have the right warning class', () => {
    const wrapper = shallowMount(Alert, {
      propsData: {
        type: 'warning',
        title: 'Hello',
        message: 'Word'
      }
    })

    expect(wrapper.classes()).toEqual([
      'border-t-4',
      'rounded-b',
      'px-4',
      'py-3',
      'shadow-md',
      'bg-orange-100',
      'border-orange-500',
      'text-orange-900'
    ])
  })
})
