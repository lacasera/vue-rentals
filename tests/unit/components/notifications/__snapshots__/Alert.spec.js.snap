// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render properly 1`] = `
<div role="alert" class="border-t-4 rounded-b px-4 py-3 shadow-md bg-red-100 border-red-500 text-red-900">
  <div class="flex">
    <div class="py-1"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" class="fill-current h-6 w-6 text-red-500 mr-4 text-red-500">
        <path d="M2.93 17.07A10 10 0 1 1 17.07 2.93 10 10 0 0 1 2.93 17.07zm12.73-1.41A8 8 0 1 0 4.34 4.34a8 8 0 0 0 11.32 11.32zM9 11V9h2v6H9v-4zm0-6h2v2H9V5z"></path>
      </svg></div>
    <div>
      <p id="alert-title" class="font-bold"> Hello </p>
      <p id="alert-message" class="text-sm"> Word</p>
    </div>
  </div>
</div>
`;
